### PR TITLE
Update quantization docs

### PR DESCRIPTION
**Pull request by DerekLiu35**
Created draft of quantization concept guide

@SunMarc

### CODE REVIEW

File: docs/source/en/quantization/concept_guide.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+$$
+x \approx S \cdot (q - Z)
+$$
+
+<img width="974" alt="dequant" src="https://gist.github.com/user-attachments/assets/05c10fc0-ace3-40ff-962c-c4ea0eb49d60" />
+
+During inference, computations like matrix multiplication are performed using the `int8` values ($q$), and the result is then dequantized back to `float32` (often using a higher-precision accumulation type like `int32` internally) before being passed to the next layer.
+
+### Symmetric vs. Asymmetric Quantization
```
Comment (MekkCyber) [inline docs/source/en/quantization/concept_guide.md:52]: I think it would be better to define Symmetric vs Asymmetric before Affine quantization

File: docs/source/en/quantization/concept_guide.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+The 🤗 Transformers library integrates with several quantization backends like:
+
+- `bitsandbytes`
+- `auto-gptq`
+- `awq`
+- And more...
+
+These are unified under the `HfQuantizer` API and associated `QuantizationConfig` classes.
+
```
Comment (MekkCyber) [inline docs/source/en/quantization/concept_guide.md:87]: we can add that the user can use his custom quantization backend wdyt ?

File: docs/source/en/quantization/concept_guide.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+### Typical Workflow
+
+1. Choose a quantization method suitable for your hardware and use-case (see the [Overview table](./overview)).
+2. Load a pre-quantized model from the Hugging Face Hub, or loading a `float32`/`float16` model and applying quantization using a specific `QuantizationConfig`.
```
Comment (MekkCyber) [inline docs/source/en/quantization/concept_guide.md:91]:
```suggestion
2. Load a pre-quantized model from the Hugging Face Hub, or loading a `float32`/`float16`/`bfloat16` model and applying quantization using a specific `QuantizationConfig`.
```

File: docs/source/en/quantization/concept_guide.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
+# Quantization Concepts
+
+Quantization is a technique used to reduce the memory footprint and computational cost of machine learning models, particularly large ones like those found in the 🤗 Transformers library. It achieves this by representing the model's weights and/or activations using lower-precision data types (like 8-bit integers - `int8`) instead of the standard 32-bit floating-point (`float32`).
+
+## Why Quantize?
+
+Reducing the precision of a model offers several significant advantages:
+
+1.  **Reduced Model Size:** Lower-precision data types require less storage space. An `int8` model, for example, is roughly 4 times smaller than its `float32` counterpart.
+2.  **Faster Inference:** Operations on lower-precision data types, especially integers, can be significantly faster on compatible hardware (CPUs and GPUs often have specialized instructions for `int8` operations). This leads to lower latency.
```
Comment (MekkCyber) [inline docs/source/en/quantization/concept_guide.md:10]: Throughout the document, there’s no mention of int4 quantization I think. It would be helpful to include a section on int4, with a brief explanation of how weights are packed and why this format is beneficial — especially in memory-bound scenarios where reduced memory bandwidth can lead to meaningful performance improvements, even without native int4 compute support.

File: docs/source/en/quantization/concept_guide.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+)
+
+model = AutoModelForCausalLM.from_pretrained(
+    model_id,
+    quantization_config=quantization_config,
+    device_map="auto"
+)
+```
+
+## Further Learning
```
Comment (MekkCyber) [inline docs/source/en/quantization/concept_guide.md:110]: not sure if we should have a section about fp8 now that it's becoming a standard, wdyt @DerekLiu35 @SunMarc

File: docs/source/en/quantization/concept_guide.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+To explore quantization more deeply, check out these short courses created in partnership with DeepLearning.AI:
+
+- [Quantization Fundamentals with Hugging Face](https://www.deeplearning.ai/short-courses/quantization-fundamentals-with-hugging-face/)
+- [Quantization in Depth](https://www.deeplearning.ai/short-courses/quantization-in-depth)
```
Comment (MekkCyber) [inline docs/source/en/quantization/concept_guide.md:115]: we can add the spaces after this section too

**Review by MekkCyber — APPROVED** [review]
Thanks a lot for the cleanup and consize doc 🔥 LGTM ! just left few nits to improve. Happy to hear you thoughts 🤗

File: docs/source/en/quantization/concept_guide.md (hunk 2 above)
Comment (DerekLiu35) [reply docs/source/en/quantization/concept_guide.md:87]: Sounds good to me

File: docs/source/en/quantization/concept_guide.md (hunk 1 above)
Comment (DerekLiu35) [reply docs/source/en/quantization/concept_guide.md]: Tried moving Symmetric vs Asymmetric before Affine quantization but not sure if my explanations still make sense

File: docs/source/en/quantization/concept_guide.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+## Why Quantize?
+
+Reducing the precision of a model offers several significant advantages:
+
+1.  **Reduced Model Size:** Lower-precision data types require less storage space. An `int8` model, for example, is roughly 4 times smaller than its `float32` counterpart.
+2.  **Faster Inference:** Operations on lower-precision data types, especially integers, can be significantly faster on compatible hardware (CPUs and GPUs often have specialized instructions for `int8` operations). This leads to lower latency.
+4.  **Reduced Energy Consumption:** Faster computations and smaller memory transfers often translate to lower power usage.
+
+## How Quantization Works (Focus on `int8`)
```
Comment (MekkCyber) [inline docs/source/en/quantization/concept_guide.md:13]:
```suggestion
## How Quantization Works
```
we talk about int4 in the same section

File: docs/source/en/quantization/concept_guide.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+A key aspect of `int4` quantization is **weight packing**. Since most hardware doesn't natively handle 4-bit data types in memory, two `int4` values are typically packed together into a single `int8` byte for storage and transfer. For example, the first value might occupy the lower 4 bits and the second value the upper 4 bits of the byte (e.g., `packed_byte = (val1 & 0x0F) | (val2 << 4)`).
+
+**Why is `int4` beneficial even without native `int4` compute?**
+
+The primary benefit comes from reduced memory bandwidth. Loading the packed `int4` weights (stored as `int8`) from memory (RAM or VRAM) to the compute units is twice as fast as loading `int8` weights. For large models, memory access is often a significant bottleneck. The speed gain from faster data transfer can outweigh the computational overhead of unpacking and dequantizing on the fly, leading to overall faster inference, especially in memory-bound scenarios.
+
+However, `int4` quantization typically results in a larger accuracy drop compared to `int8`. Advanced quantization techniques like GPTQ or AWQ are often necessary to achieve good performance with `int4`.
```
Comment (MekkCyber) [inline docs/source/en/quantization/concept_guide.md:76]: Very nice 🤗

File: docs/source/en/quantization/concept_guide.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+## Quantization in 🤗 Transformers
+
+The 🤗 Transformers library integrates with several quantization backends like:
+
+- `bitsandbytes`
+- `auto-gptq`
+- `awq`
+- And more...
+
+These are unified under the `HfQuantizer` API and associated `QuantizationConfig` classes. You can integrate your own custom quantization backends by implementing a custom `HfQuantizer` and `QuantizationConfig`, allowing seamless use within the 🤗 Transformers ecosystem.
```
Comment (MekkCyber) [inline docs/source/en/quantization/concept_guide.md:107]:
```suggestion
These are unified under the `HfQuantizer` API and associated `QuantizationConfig` classes. You can integrate your own custom quantization backends by implementing a custom `HfQuantizer` and `QuantizationConfig`, allowing seamless use within the 🤗 Transformers ecosystem, for more details you can check out this [example](https://github.com/huggingface/transformers/blob/main/examples/quantization/custom_quantization_int8_example.py).
```

**Review by MekkCyber — APPROVED** [review]
Thanks for the new sections @DerekLiu35 ! Awesome read ! left only few nits

File: docs/source/en/quantization/concept_guide.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    device_map="auto"
+)
+```
+
+## Further Learning
+
+To explore quantization more deeply, check out these short courses created in partnership with DeepLearning.AI:
+
+- [Quantization Fundamentals with Hugging Face](https://www.deeplearning.ai/short-courses/quantization-fundamentals-with-hugging-face/)
+- [Quantization in Depth](https://www.deeplearning.ai/short-courses/quantization-in-depth)
```
Comment (SunMarc) [inline docs/source/en/quantization/concept_guide.md:136]: we can also add https://horace.io/brrr_intro.html
or 
https://web.archive.org/web/20241004103940/https://pytorch.org/blog/accelerating-generative-ai-2/ -> need to fix the link 
to show what you mean by memory / compute bound 

Also I would recommend the MIT han lab class ressources for deeper understanding on quantizaiton. 

Merve also have a blogpost around quantization https://huggingface.co/blog/merve/quantization

File: docs/source/en/quantization/concept_guide.md [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+### Typical Workflow
+
+1. Choose a quantization method suitable for your hardware and use-case (see the [Overview table](./overview)).
+2. Load a pre-quantized model from the Hugging Face Hub, or loading a `float32`/`float16`/`bfloat16` model and applying quantization using a specific `QuantizationConfig`.
+
+```python
+from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig
+
+model_id = "facebook/opt-350m"
```
Comment (SunMarc) [inline docs/source/en/quantization/concept_guide.md:117]: maybe select a newer model like a llama or a smollm?

File: docs/source/en/quantization/concept_guide.md [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+- `auto-gptq`
+- `awq`
+- And more...
+
+These are unified under the `HfQuantizer` API and associated `QuantizationConfig` classes. You can integrate your own custom quantization backends by implementing a custom `HfQuantizer` and `QuantizationConfig`, allowing seamless use within the 🤗 Transformers ecosystem, for more details you can check out this [example](https://github.com/huggingface/transformers/blob/main/examples/quantization/custom_quantization_int8_example.py).
+
+### Typical Workflow
+
+1. Choose a quantization method suitable for your hardware and use-case (see the [Overview table](./overview)).
+2. Load a pre-quantized model from the Hugging Face Hub, or loading a `float32`/`float16`/`bfloat16` model and applying quantization using a specific `QuantizationConfig`.
```
Comment (SunMarc) [inline docs/source/en/quantization/concept_guide.md:112]: you can also add a reference to the selecting a quantization method docs that you wrote

File: docs/source/en/quantization/concept_guide.md [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+quantization_config = BitsAndBytesConfig(
+    load_in_4bit=True,
+    bnb_4bit_compute_dtype=torch.bfloat16
+)
+
+model = AutoModelForCausalLM.from_pretrained(
+    model_id,
+    quantization_config=quantization_config,
+    device_map="auto"
+)
```
Comment (SunMarc) [inline docs/source/en/quantization/concept_guide.md:128]: let's load the model with `torch_dtype = torch.bfloat16`

File: docs/source/en/quantization/concept_guide.md (hunk 5 above)
Comment (SunMarc) [reply docs/source/en/quantization/concept_guide.md:131]: A small section dedicated to fp8 could make sense indeed -> talk about A8W8 scheme in particular.

File: docs/source/en/quantization/concept_guide.md [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+The primary trade-off in quantization is **efficiency vs. accuracy**. Reducing precision saves resources but inevitably introduces small errors (quantization noise). The goal is to minimize this error using appropriate schemes (affine/symmetric), granularity (per-tensor/channel), and techniques (PTQ/QAT) so that the model's performance on its target task degrades as little as possible.
+
+## Quantization in 🤗 Transformers
+
+The 🤗 Transformers library integrates with several quantization backends like:
+
+- `bitsandbytes`
+- `auto-gptq`
+- `awq`
+- And more...
```
Comment (SunMarc) [inline docs/source/en/quantization/concept_guide.md:105]:
```suggestion
- `bitsandbytes`
-`torchao`
-`compressed-tensors`
- `awq`
- `gptqmodel`
- And more...
```

File: docs/source/en/quantization/selecting.md [hunk 15]
```diff
@@ -130,6 +130,27 @@ Methods like [AQLM](./aqlm), [SpQR](./spqr), [VPTQ](./vptq), [HIGGS](./higgs), e
...
+*Results for `torch.compile` and `Marlin` kernels are included where applicable and supported.*
+
+The full dataset with detailed results can be found on the Hugging Face Hub: 
+<iframe
+  src="https://huggingface.co/datasets/derekl35/quantization-benchmarks/embed/viewer/default/train"
+  frameborder="0"
+  width="100%"
+  height="560px"
+  title="benchmarking results dataset"
+></iframe>
```
Comment (SunMarc) [inline docs/source/en/quantization/selecting.md:147]: do you have the benchmarking script ? Might be good to link it

**Review by SunMarc — APPROVED** [review]
Thanks for adding this ! Love it

File: docs/source/en/quantization/selecting.md [hunk 16]
```diff
@@ -130,6 +130,27 @@ Methods like [AQLM](./aqlm), [SpQR](./spqr), [VPTQ](./vptq), [HIGGS](./higgs), e
     *   You have significant compute resources available for potentially complex quantization procedures.
 We recommend consulting each methods documentation and associated papers carefully before choosing one for use in production.
 
+## Benchmark Comparison
+
+To provide a quantitative comparison of different quantization methods, we benchmarked several popular techniques on the Llama 3.1 8B and 70B models. The following tables show results for accuracy (higher is better), inference throughput (tokens/second, higher is better), peak VRAM usage (GB, lower is better), and quantization time.
+
+*Performance metrics were measured on 2 NVIDIA A100 80GB GPU for Llama 3.1 70B - bfloat16, 1 NVIDIA h100 80GB GPU for FP8 methods, and 1 NVIDIA A100 80GB GPU for all other methods. Throughput was measured with a batch size of 1 and generating 64 tokens.*
+*Results for `torch.compile` and `Marlin` kernels are included where applicable and supported.*
```
Comment (stevhliu) [inline docs/source/en/quantization/selecting.md:138]:
```suggestion
Performance metrics were measured on 2 NVIDIA A100 80GB GPU for Llama 3.1 70B (bfloat16), 1 NVIDIA H100 80GB GPU for FP8 methods, and 1 NVIDIA A100 80GB GPU for all other methods. Throughput was measured with a batch size of 1 and generating 64 tokens.
Results for `torch.compile` and Marlin kernels are included where applicable and supported.
```

File: docs/source/en/quantization/selecting.md [hunk 17]
```diff
@@ -130,6 +130,27 @@ Methods like [AQLM](./aqlm), [SpQR](./spqr), [VPTQ](./vptq), [HIGGS](./higgs), e
...
+
+The full dataset with detailed results can be found on the Hugging Face Hub:
```
Comment (stevhliu) [inline docs/source/en/quantization/selecting.md:140]:
```suggestion
```

**Review by stevhliu — COMMENTED** [review]
Wow this is a really nice resource, great work! 👏

File: docs/source/en/quantization/selecting.md (hunk 15 above)
Comment (DerekLiu35) [reply docs/source/en/quantization/selecting.md]: I have some benchmarking code but the code (https://github.com/DerekLiu35/huggingface-quantization-benchmarking-code/tree/main) isn't that clean
